experimental/unicodeset_parser/src/parse.rs

```diff
@@ -0,0 +1,1192 @@
+                    self.iter.next();
+                    break;
+                }
+                // parse the property name
+                (Begin | PropertyName, c) if c.is_ascii_alphanumeric() || c == '_' => {
+                    key_buffer.push(c);
+                    self.iter.next();
+                    state = PropertyName;
+                }
+                // parse the name-value separator
+                (PropertyName, c @ ('=' | '≠')) => {
+                    equality = c == '=';
+                    self.iter.next();
+                    state = PropertyValueBegin;
+                }
+                // parse the property value
+                (PropertyValue | PropertyValueBegin, c) if c != end => {
+                    value_buffer.push(c);
+                    self.iter.next();
+                    state = PropertyValue;
+                }
+                (_, c) => return self.error_here(PEK::UnexpectedChar(c)),
+            }
+        }
+
+        if !equality {
+            self.inverted = !self.inverted;
+        }
+
+        let (set, inverted) = self
+            .load_property_codepoints(&key_buffer, &value_buffer)
+            // any error that does not already have an offset should use the appropriate property offset
+            .map_err(|e| e.or_with_offset(property_offset))?;
+        if inverted {
+            self.inverted = !self.inverted;
+        }
+        self.single_set.add_set(&set.to_code_point_inversion_list());
+
+        Ok(())
+    }
+
+    // also returns whether the set needs to be inverted or not
+    fn load_property_codepoints(&self, key: &str, value: &str) -> Result<(CodePointSetData, bool)> {
+        // we support:
+        // [:gc = value:]
+        // [:sc = value:]
+        // [:scx = value:] -- TODO: implement scx
+        // [:value:] - looks up value in gc, sc and scx
+        // [:prop:] - binary property, returns codepoints that have the property
+        // [:prop = truthy/falsy:] - same as above
+
+        let mut inverted = false;
+
+        if value.is_empty() {
+            // key is binary property, or a value of gc, sc, scx
+
+            // try loading a binary property, if it fails, try loading a gc, sc, scx value
+            let set = load_for_ecma262_unstable(self.property_provider, key)
+                .or_else(|_| self.try_load_general_category(key))
+                .or_else(|_| self.try_load_script(key))?;
+            // .or_else(|_| self.try_load_script_extensions(key))?;
+
+            Ok((set, inverted))
+        } else {
+            // key is gc, sc, scx
+            // value is a property value
+            // OR
+            // key is a binary property and value is a truthy/falsy value
+
+            // UnicodeProperty::parse_ecma262_name would be good to have to avoid this duplication:
+            let set = match key {
+                "General_Category" | "gc" => self.try_load_general_category(value)?,
```

Comment:
Suggestion (optional): It would be a little bit nicer if you could write this function with a single invocation of each of the three load functions. With multiple call sites, the functions are less likely to be inlined (and if they are, they increase code size).

Reply:
I can't use something like `opt1.iter().chain(opt2.iter()).chain(opt3.iter())`, because I need to know which property-loading function to call (and that would mean I'd have to put some identifier in the Some(), resulting in questionable style improvements)

Reply:
I like the Results version